driver/src/main/com/mongodb/client/MongoClient.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2017-2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import com.mongodb.ClientSessionOptions;
+import com.mongodb.MongoNamespace;
+import com.mongodb.annotations.Immutable;
+import com.mongodb.session.ClientSession;
+import org.bson.Document;
+
+/**
+ * A client-side representation of a MongoDB cluster.  Instances can represent either a standalone MongoDB instance, a replica set,
+ * or a sharded cluster.  Instance of this class are responsible for maintaining an up-to-date state of the cluster,
+ * and possibly cache resources related to this, including background threads for monitoring, and connection pools.
+ * <p>
+ * Instances of this class serve as factories for {@link MongoDatabase} instances.
+ * </p>
+ * <p>
+ * Instances of this class can be created via the {@link MongoClients} factory.
+ * </p>
+ * @see MongoClients
+ * @since 3.7
+ */
+@Immutable
+public interface MongoClient {
```

Comment:
com.mongodb.MongoClient could implement this interface without any change, but it may be awkward for users  for this class to implement an interface with the same simple name.